### PR TITLE
[feat] - use diff chan

### PR DESCRIPTION
**Pull request by ahrav**
### Description:
Replaces the commit chan with a diff chan.

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

### CODE REVIEW

File: pkg/gitparse/gitparse.go
Comment (ahrav) [inline pkg/gitparse/gitparse.go:159]: This was only used for testing so I moved it to the test file.

File: pkg/gitparse/gitparse_test.go
Comment (ahrav) [inline pkg/gitparse/gitparse_test.go:573]: This was removed as we do not use the commit message at any point later in the pipeline. If we need it in the future we can always add it back.

Comment (rgmz) [reply pkg/gitparse/gitparse_test.go:573]: Do we scan commit messages? If not, we should.

Comment (ahrav) [reply pkg/gitparse/gitparse_test.go:573]: We do not currently, @zricethezav also mentioned there were plans to, but nothing definitive that i'm aware of.

Comment (rgmz) [reply pkg/gitparse/gitparse_test.go:573]: It's quite simple to implement with the current design, so I think it's worth considering how it could be done with this new implementation.

Comment (ahrav) [reply pkg/gitparse/gitparse_test.go:573]: Yea, I don't think it should be too much more complicated than with the prior implementation. (at least i hope not 😅 )

File: pkg/sources/git/git.go
Comment (bill-rich) [inline pkg/sources/git/git.go:558]: I think this should only be logged when moving onto a new commit.

File: pkg/gitparse/gitparse.go
Comment (bill-rich) [inline pkg/gitparse/gitparse.go:109]: What about keeping these in `Commit` and just having a pointer in `Diff`? That way we don't need to copy all this data for each `Diff`.

Comment (ahrav) [reply pkg/gitparse/gitparse.go:109]: I defs thought about that, and now that you bring it up, I think i'll go that route.

File: pkg/sources/git/git.go
Comment (ahrav) [reply pkg/sources/git/git.go:558]: Good call, yep!

File: pkg/gitparse/gitparse.go
Comment (ahrav) [reply pkg/gitparse/gitparse.go:109]: Well actually now that i'm looking at it again, we don't even need the author or date in the commit and longer. We only need the hash to stay in the commit. Is it worth adding a ptr to the commit for just the hash?

File: pkg/sources/git/git.go
Comment (mcastorina) [inline pkg/sources/git/git.go:554]: This relies on `diffChan` to send diffs in order.

I think that is the case currently, but if we ever add concurrency it won't hold. Maybe we can add a doc comment somewhere (`RepoPath` ?) that notes this behavior.

File: pkg/sources/git/git.go
Comment (mcastorina) [inline pkg/sources/git/git.go:556]: Is `s.metrics.commitsScanned` shared with other goroutines?

Comment (ahrav) [reply pkg/sources/git/git.go:556]: Yea i think it is... i'm not entirely sure so I didn't change the exiting behavior jut to be safe.

File: pkg/sources/git/git.go
Comment (ahrav) [reply pkg/sources/git/git.go:554]: Yea will do.

**Review by mcastorina — CHANGES_REQUESTED** [review]
I did a quick smoke test with your changes, and the most immediate (and hopefully only) regression I saw was the timestamps were the null time value in the found results:

```
» ./trufflehog git --no-update https://github.com/trufflesecurity/test_keys.git
```

```
...
Commit: 77b2a3e56973785a52ba4ae4b8dac61d4bac016f
File: keys
Line: 6
Repository: https://github.com/trufflesecurity/test_keys.git
Timestamp: 0001-01-01 00:00:00 +0000
           ^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: pkg/gitparse/gitparse.go
Comment (ahrav) [reply pkg/gitparse/gitparse.go]: nvm, using the commit is a good idea after all. 🤣

**Comment by ahrav** [comment]
> I did a quick smoke test with your changes, and the most immediate (and hopefully only) regression I saw was the timestamps were the null time value in the found results:
> 
> ```
> » ./trufflehog git --no-update https://github.com/trufflesecurity/test_keys.git
> ```
> 
> ```
> ...
> Commit: 77b2a3e56973785a52ba4ae4b8dac61d4bac016f
> File: keys
> Line: 6
> Repository: https://github.com/trufflesecurity/test_keys.git
> Timestamp: 0001-01-01 00:00:00 +0000
>            ^^^^^^^^^^^^^^^^^^^^^^^^^
> ```

should be fixed now, feel free to try again.
Screenshot 2024-02-05 at 8 52 02 PM

**Review by mcastorina — APPROVED** [review]
Nothing really jumps out at me as incorrect, and the smoke test output matches what we have in main.

I'm approving, but I would feel better about merging if you got another approval from someone more experienced with this package.

**Comment by ahrav** [comment]
> Nothing really jumps out at me as incorrect, and the smoke test output matches what we have in main.
> 
> 
> 
> I'm approving, but I would feel better about merging if you got another approval from someone more experienced with this package.

Thanks Miccah, I'm gonna let it get some more eyes tomorrow. I also have a couple more tests I'd like to work through just to be safe.

File: pkg/gitparse/gitparse.go
Comment (zricethezav) [inline pkg/gitparse/gitparse.go:437]: Any reason we can't do this? (suggestion, feel free to ignore)
```
if currentCommit == nil {
   currentDiff.Commit = &Commit{}
 }
```

File: pkg/gitparse/gitparse.go
Comment (zricethezav) [inline pkg/gitparse/gitparse.go:522]: Are there any concerns with dropping all these `isStaged`? Don't we use that for pre-commit scanning?

File: pkg/gitparse/gitparse_test.go
Comment (zricethezav) [reply pkg/gitparse/gitparse_test.go:573]: > It's quite simple to implement with the current design, so I think it's worth considering how it could be done with this new implementation.

Sounds like another issue and implementation PR. We should wait before this PR is merged before considering it.

File: pkg/gitparse/gitparse.go
Comment (ahrav) [reply pkg/gitparse/gitparse.go:522]: The only reason I dropped them was because my linter was yelling at me that they were unused. I couldn't find any references to them, but i'm happy to add them back in case I overlooked something. I probably should've made this change in a separate PR, but I figured I was already cleaning this file up that i'd tack it on 😓

**Review by rosecodym — APPROVED** [review]
This looks really nice, man! I also don't feel like enough of an expert on the package specifically to give my approval much weight but the code changes themselves look really straightforward.

File: pkg/gitparse/gitparse.go
Comment (zricethezav) [reply pkg/gitparse/gitparse.go:522]: I'd trust the linter, s'all good 👍🏻

**Review by zricethezav — APPROVED** [review]

**Review by bill-rich — APPROVED** [review]
Looks good! Excited to see these changes in action.

File: pkg/gitparse/gitparse.go
Comment (ahrav) [reply pkg/gitparse/gitparse.go:309]: I'm actually not sure... I think @bill-rich might have some insight into that value being chosen.

File: pkg/gitparse/gitparse.go
Comment (ahrav) [reply pkg/gitparse/gitparse.go:830]: You ask good questions.... of which I don't know the answer too. @bill-rich could i bug you again for this one. 🙇

**Pull request merged**
